File: problem2.java
// Problem2 - https://leetcode.com/problems/01-matrix/

// Time Complexity : O(m*n)
// Space Complexity : O(m*n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : Nothing as such

class BFS {
    public int[][] updateMatrix(int[][] mat) {
        if (mat == null || mat.length == 0)
            return mat;
        // Updating Matrix with all 1 into -1
        int m = mat.length;
        int n = mat[0].length;
        Queue<int[]> q = new LinkedList<>();
        int[][] dirs = new int[][] { { 0, -1 }, { 1, 0 }, { 0, 1 }, { -1, 0 } };

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (mat[i][j] == 1) {
                    mat[i][j] = -1;
                }
            }
        }

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (mat[i][j] == 0) {
                    q.add(new int[] { i, j });
                }
            }
        }

        int dist = 0;
        while (!q.isEmpty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                int[] curr = q.poll();
                for (int[] dir : dirs) {
                    int nr = dir[0] + curr[0];
                    int nc = dir[1] + curr[1];

                    // bounds check
                    if (nr >= 0 && nc >= 0 && nr < m && nc < n && mat[nr][nc] == -1) {
                        mat[nr][nc] = dist + 1;
                        q.add(new int[] { nr, nc });
                    }
                }
            }
            dist++;
        }
        return mat;
    }
}

File: Problem_2.js
// Problem2 (https://leetcode.com/problems/01-matrix/)
// Time Complexity : O(mxn)
// Space Complexity : O(mxn)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

/**
 * @param {number[][]} mat
 * @return {number[][]}
 */
var updateMatrix = function (mat) {
    if (mat === null || mat.length === 0)
        return [];

    let q = [];
    let dirs = [[0, 1], [1, 0], [-1, 0], [0, -1]];
    let rows = mat.length;
    let cols = mat[0].length;

    // Add all the 0's location in queue as it is independent
    // Mark the 1's as -1 to make sure they are not visited yet
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            if (mat[i][j] === 0)
                q.push([i, j]);
            else
                mat[i][j] = -1;
        }
    }

    while (q.length > 0) {
        let node = q.shift();
        dirs.forEach(dir => {
            // For each direction if the cell is valid and it is -1 i.e not visited, set its value to  node location+1
            // Push it to queue
            let nr = node[0] + dir[0];
            let nc = node[1] + dir[1];
            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && mat[nr][nc] === -1) {
                mat[nr][nc] = mat[node[0]][node[1]] + 1;
                q.push([nr, nc]);
            }
        })
    }
    return mat;
};